### PR TITLE
[SPARK-26222][SQL] Track file listing time

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

Tracking file listing time and add them into scan node's metrics, also add the start and end timestamp into the metadata message of the plan.
image

## How was this patch tested?

Add new tests in SQLMetricsSuite.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -313,8 +321,10 @@ case class FileSourceScanExec(
   override lazy val metrics =
     Map("numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
       "numFiles" -> SQLMetrics.createMetric(sparkContext, "number of files"),
+      "scanTime" -> SQLMetrics.createTimingMetric(sparkContext, "scan time"),
       "fileListingTime" -> SQLMetrics.createMetric(sparkContext, "file listing time (ms)"),
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Currently, the file listing time still contains part of partition pruning time(introduced from [SPARK-20151](https://github.com/apache/spark/pull/17476/files)), I think we should split the partition pruning time into `metastore operation time`, which will be done in https://issues.apache.org/jira/browse/SPARK-26223.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 2]
```diff
@@ -167,28 +167,30 @@ case class FileSourceScanExec(
...
    * been initialized. See SPARK-26327 for more details.
    */
   private def sendDriverMetrics(): Unit = {
+    val driverMetrics: HashMap[String, Long] = HashMap.empty
+
+    driverMetrics("numFiles") = selectedPartitions.map(_.files.size.toLong).sum
+    val fileListingPhase = relation.location.fileListingPhase
+    if (fileListingPhase.isDefined) {
+      val phase = fileListingPhase.get
+      driverMetrics("metadataTime") = phase.durationMs
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: As discussion with @gatorsmile, `metadataTime` maybe still a clearer name for users, change it back in b79e948.
As my current plan, we'll finally get five metrics as below:
"metadata time": including all metadata accessing time like file listing time, meta store time and others.
"file listing start/end": the start and end timestamp for file listing
"metastore operation start/end": the start and end timestamp for metastore operation.

**Comment by xuanyuanking** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 3]
```diff
@@ -325,8 +327,10 @@ case class FileSourceScanExec(
   override lazy val metrics =
     Map("numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
       "numFiles" -> SQLMetrics.createMetric(sparkContext, "number of files"),
-      "metadataTime" -> SQLMetrics.createMetric(sparkContext, "metadata time"),
-      "scanTime" -> SQLMetrics.createTimingMetric(sparkContext, "scan time"))
+      "scanTime" -> SQLMetrics.createTimingMetric(sparkContext, "scan time"),
+      "metadataTime" -> SQLMetrics.createMetric(sparkContext, "metadata time (ms)"),
+      "fileListingStart" -> SQLMetrics.createTimestampMetric(sparkContext, "file listing start"),
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Here's a problem existing, for the metrics like metadataTime and start/end timestamp, they are driver-only, which only create, update and display controlled on driver-side. The current usage has little waste on sending metrics to the executor and aggregates back. I'll address this in a separate JIRA and PR, this is a common work for both SPARK-26222/SPARK-26223.

**Comment by rxin** [comment]
Can we change the output in UI to show only the time duration, and when hover show both start and end time? The actual underlying recorded metrics should only include start / end time.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 4]
```diff
@@ -131,6 +133,14 @@ object SQLMetrics {
     acc
   }
 
+  def createTimestampMetric(sc: SparkContext, name: String): SQLMetric = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: createTimelineMetric

**Review by rxin — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala [hunk 5]
```diff
@@ -87,6 +87,15 @@ object QueryPlanningTracker {
     localTracker.set(tracker)
     try f finally { localTracker.set(originalTracker) }
   }
+
+  /** Record the phase by the passing in record function. */
+  def recordPhase[T](record: PhaseSummary => Unit)(f: => T): T = {
```
[the remark this answers is not in the document]
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala]: Copy, I simply this by returning tuple of both result and phase summary.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala [hunk 6]
```diff
@@ -74,12 +74,11 @@ trait FileIndex {
...
    *
-   * We do file listing in query optimization (in order to get the proper statistics) and we want
-   * to account for file listing time in physical execution (as metrics). To do that, we save the
-   * file listing time in some implementations and physical execution calls it in this method
-   * to update the metrics.
+   * We call this in physical execution while we want to account for the file listing time as
+   * metrics. If partition pruning happened in query planning, the phase also contains this
+   * part of the cost, otherwise, it only contains file listing time of FileIndex initialize.
    */
-  def metadataOpsTimeNs: Option[Long] = None
```
[the remark this answers is not in the document]
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala]: I try to do like this but the `metadataOpsTimeNs` is part of whole metadata time, valued it by CatalogFileIndex's fileListingPhase.durationMs would get a wrong answer. Worrying about not knowing your meaning precisely, I just keep this in next commit.
I plan to address this as part of the metastore time in next JIRA SPARK-26223, also summary all metastore time and file listing time as final metadata time, do you agree about this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala [hunk 7]
```diff
@@ -74,12 +74,11 @@ trait FileIndex {
...
+  def fileListingPhase: Option[QueryPlanningTracker.PhaseSummary] = None
```
[the remark this answers is not in the document]
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala]: Copy that, done in 1ed45b1.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala (hunk 4 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: Copy, changes in SQLMetrics has been reverted in next commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala]:
```
I plan to address this as part of the metastore time in next JIRA SPARK-26223
```
Here's the part in preview PR: https://github.com/apache/spark/pull/23371/files#diff-4f50a25b08ee9a2bb1d12658ead0a455R74

**Comment by xuanyuanking** [comment]
Gentle ping for review when you have time, Thanks :) @rxin @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CatalogFileIndex.scala [hunk 8]
```diff
@@ -56,11 +60,17 @@ class CatalogFileIndex(
 
   override def listFiles(
       partitionFilters: Seq[Expression], dataFilters: Seq[Expression]): Seq[PartitionDirectory] = {
-    filterPartitions(partitionFilters).listFiles(Nil, dataFilters)
+    val (partitions, phase) = createPhaseSummary {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CatalogFileIndex.scala]: how hard it is to do it at caller side?

Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CatalogFileIndex.scala]: Thank you Wenchen for your review.
Not so hard, but need more code changes on caller class. I choose this just for little change and code clean. The current approach need call `createPhaseSummary`  here and `InmemoryFileIndex.refresh0`, if changes to caller side need tracker more place call listFiles and `new InMemoryFileIndex`.
Another problem is how to pass the phase to ScanExec. If we agree still keep it in FileIndex, need a set function of the phaseSummary for caller side.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala [hunk 9]
```diff
@@ -82,4 +83,16 @@ trait FileIndex {
...
+   * Returns the latest phase summary of file listing in the current FileIndex, we should also
+   * clean the phase summary cause in the scenario of the cached plan, we shouldn't report the
+   * old phase summary.
+   * This interface is only overridden in [[InMemoryFileIndex]] and [[CatalogFileIndex]], we do
+   * not override this in [[PartitioningAwareFileIndex]] cause all its subclass using in scan
+   * node already track file listing time.
+   *
+   * @return An optional phase summary to record the start and end timestamp for listing file.
+   */
+  def getAndCleanFileListingPhaseSummary: Option[PhaseSummary] = None
```
Comment (LantaoJin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala]: Just FYI. After patched this method, current Delta-Lake df.show will throw `AbstractMethodError`
```
java.lang.AbstractMethodError
  at org.apache.spark.sql.execution.FileSourceScanExec.fileListingPhaseSummary$lzycompute(DataSourceScanExec.scala:248)
```

**Pull request closed without merging**